File: Python/198.py
class Solution:
    # @param {integer[]} nums
    # @return {integer}
    """ Version 1.0 First try
    def rob(self, nums):
        if not nums:
            return 0
        elif len(nums) < 3:
            return max(nums)
        elif len(nums) == 3:
            return max(nums[1],nums[0]+nums[2])
        dict_m = {0:nums[0],1:nums[1],2:nums[0]+nums[2]}
        m = 0
        for i in range(len(nums))c
            m = max(dict_m[i-3] + nums[i], dict_m[i-2] + nums[i], dict_m[i-1])
            if i not in dict_m:
                dict_m[i] = m
        return m
    """ 
    """ Version 2.0 list but dict, structure！dict_m[i-2] + nums[i] >= dict_m[i-3] + nums[i]
    def rob(self, nums):
        length = len(nums)
        if length == 0:
            return 0
        elif length <3:
            return max(nums)
        m = [nums[0]]
        m.append(max(nums[0],nums[1]))
        for i in range(length)[2:]:
            m.append(max(m[i-2] + nums[i], m[i-1]))
        return m[-1]
    """
    # Version 3.0 Awesome!
    def rob(self, num):
        left = left_left = 0
        for i in xrange(len(num)):
            left, left_left =  max(num[i] + left_left, left), left
        return left
    

File: C++/198.cpp
class Solution {
public:
    /* Version 1
    int rob(vector<int>& nums) {
        if(nums.empty()) return 0;
        int prev = 0, res = nums[0]; // prev is 0;
        for(int i = 1; i < nums.size(); i++){
            if(nums[i] + prev > res) {
                int temp = res;
                res = nums[i] + prev;
                prev = temp;
            }
            else{
                prev = res;
            }
        }
        return res;
    }
    */
    
    // Consice Solution
    int rob(vector<int>& nums) {
        int prev = 0, curr = 0;
        for(int num : nums){
            int temp = max(num + prev, curr);
            prev = curr;
            curr = temp;
        }
        return curr;
    }
};